analysis/server.R

```diff
@@ -50,94 +50,107 @@ server <- function(input, output, session) {
+  
+  # Create base leaflet plot with geoJSON region boundaries
+  scot_plot_reactive <- reactive({
+    renderLeaflet({
+      # Apply default styling to all features (colours will be overridden later)
+      leaflet(scotland) %>%
+        addGeoJSONv2(
+          layerId = "healthBoardGeoJSON",
+          geojson_health_boards,
+          weight = 1,
+          color = "red",
+          fillColor = "red",
+          opacity = 1,
+          fillOpacity = 0.8,
+          highlight = highlightOptions(
+            weight = 3,
+            opacity = 0.5,
+            fillOpacity = 0.5,
+            bringToFront = TRUE
+          )
+          
+        ) %>%
+        fitBounds(-7.5, 54.6, -0.4, 61.0)
+    })
+  })
+  
+  observe({
+    output$scot_plot <- scot_plot_reactive()
+  })
+  
+  # Update palette when management variable changes
+  scot_plot_pal_reactive <- reactive({
     # creates bins and palette for leaflet plot
-    # bins <- seq(0, max(management_reactive()$total), length.out = 6)
-
-    pal <- colorBin("plasma", domain = scotland_count$value, bins = 5)
-
-    # creates hover over labels
-    labels <- sprintf(
-      "<strong>%s</strong><br/>%g",
-      scotland_count$HBName,
-      scotland_count$value
-    ) %>% lapply(htmltools::HTML)
-
-
-    scotland_count %>%
-      leaflet() %>%
-      addPolygons(
-        fillColor = ~ pal(value),
-        weight = 2,
-        opacity = 1,
-        color = "white",
-        dashArray = "3",
-        fillOpacity = 0.5,
-        highlight = highlightOptions(
-          weight = 5,
-          color = "#666",
-          dashArray = "",
-          fillOpacity = 0.7,
-          bringToFront = TRUE
-        ),
-        label = labels,
-        labelOptions = labelOptions(
-          style = list(
-            "font-weight" = "normal",
-            padding = "3px 8px"
-          ),
-          textsize = "15px",
-          direction = "auto"
-        )
-      ) %>%
+    colorBin("plasma", domain = management_by_variable_reactive()$value, bins = 5)
+  })
+  
+  # Update legend when management variable changes
+  observe({
+    leafletProxy("scot_plot", data =  scotland) %>%
       addLegend(
-        pal = pal,
-        values = ~value,
+        layerId = "managementLegend",
+        pal = scot_plot_pal_reactive(),
+        values = management_by_variable_reactive()$value,
         opacity = 0.7,
         title = "Count",
         position = "topleft"
       )
   })
-
+  
+  # Update region colours and tooltips when management variable or date change
+  observe({
+    pal <- scot_plot_pal_reactive()
+    featureData <-
+      management_by_variable_and_date_reactive() %>% select(areaname, value)
+    featureData$colour <- pal(featureData$value)
+    
+    leafletProxy("scot_plot", data =  scotland) %>%
+      leaflet::invokeMethod(
+        scotland,
+        "setRegionStyleAndTooltips",
+        "healthBoardGeoJSON",
+        featureData,
+        pal(NA)
+      )
+  })
```

Comment:
Pretty sure I know what's happening here but I may need you to explain how 👍

Reply:
This invokes the client JS function setRegionStyleAndTooltips (in ui.R), supplying a subset of the columns from management (after filtering by variable and date). The original leaflet plot with the geoJSON boundaries is reused, so no time spent rebuilding it. The JS function updates the leaflet layer in place.